[Service Bus] Pick TestClientType randomly

ramya-rao-a opened the pull request:
Currently we have quite a few tests in Service Bus that are repeated for 8 different cases
- Partitioned Queue
- UnPartitioned Queue
- Partitioned Queue with Sessions
- UnPartitioned Queue with Sessions
- Partitioned Subscription
- UnPartitioned Subscription
- Partitioned Queue with Subscription
- UnPartitioned Queue with Subscription

The reasons:
- Sessions vs No sessions variation because they follow different code paths
- Queues vs Subscriptions made sense in Track 1, but not Track 2, as we had different code paths due to the client hierarchy which has now changed
- Partitioned vs UnPartitioned was added to the mix because we observed that the service behaved differently in certain cases due to code paths being different on the service side. While most of these are fixed on the service side, we retained the tests to catch future regressions.

This PR proposes a change such that we
- retain the session vs no session variations
- but randomly choose between queue vs subscription and partition vs unpartition

This way, we get the coverage we need, but reduce the number of tests in each run, thus reducing the total time taken which at the moment is over 2 hours.

cc @HarshaNalluru, @richardpark-msft, @chradek   

PS: Use the `Hide whitespace changes` option when reviewing the diff

HarshaNalluru commented on sdk/servicebus/service-bus/test/utils/testutils2.ts:
```diff
@@ -132,6 +132,18 @@ export async function drainAllMessages(receiver: Receiver<{}>): Promise<void> {
+ * Returns a TestClientType for either a Queue or a Subscription
+ * @param useSessions 
+ */
+export  function getRandomeReceiverTestClientType(useSessions: boolean): TestClientType {
```
```suggestion
export  function getRandomReceiverTestClientType(useSessions: boolean): TestClientType {
```

ramya-rao-a replied:
Fixed with https://github.com/Azure/azure-sdk-for-js/pull/9950/commits/08e64917500aaf5ce634bec7f133caec860f786f

HarshaNalluru commented on sdk/servicebus/service-bus/test/backupMessageSettlement.spec.ts:
```diff
@@ -45,117 +48,60 @@ describe("Backup message settlement - Through ManagementLink", () => {
+    return msgs[0];
+  }
+
+  [noSessionTestClientType, withSessionTestClientType].forEach((testClientType) => {
+    it(testClientType + ": complete() removes message", async function(): Promise<void> {
```
How about bringing all the `it` blocks here under one `describe` block and wrapping this 
`[noSessionTestClientType, withSessionTestClientType].forEach((testClientType) => {` 
around that `describe` block so that we don't have to repeat this for every test?

ramya-rao-a replied:
That is a good idea!

Done with https://github.com/Azure/azure-sdk-for-js/pull/9950/commits/66682d5abe7e3d15762ce2a3a2cb0ca49b9a31fd

This is how the results would look like

<img width="536" alt="Screen Shot 2020-07-08 at 12 45 16 PM" src="https://user-images.githubusercontent.com/16890566/86963140-e6932400-c118-11ea-87d0-8ac63de86abc.png">

HarshaNalluru commented:
Loving this PR :)

richardpark-msft commented:
I wonder if we could say: CI runs do this random behavior _but_ nightly CI runs do the full tests? Is that a possibility?

richardpark-msft commented on sdk/servicebus/service-bus/test/utils/testutils2.ts:
```diff
@@ -132,6 +132,18 @@ export async function drainAllMessages(receiver: Receiver<{}>): Promise<void> {
+  const noSessionRecieverClientTypes = [TestClientType.PartitionedQueue, TestClientType.PartitionedSubscription, TestClientType.UnpartitionedQueue, TestClientType.UnpartitionedSubscription];
+  const withSessionRecieverClientTypes = [TestClientType.PartitionedQueueWithSessions, TestClientType.PartitionedSubscriptionWithSessions, TestClientType.UnpartitionedQueueWithSessions, TestClientType.UnpartitionedSubscriptionWithSessions];
+
+  const index = Math.floor(Math.random() * noSessionRecieverClientTypes.length);
```
Rather than being totally random can we base it on something that's a bit more stable? Otherwise it'll be difficult to chase CI failures if we get it only during a specific branch (unlikely, but it'd be nice to never have that issue).

Perhaps we could base it on the day or something of that nature?

ramya-rao-a replied:
I made sure that the test title reflect the TestClientType. This is exactly for the case you mentioned i.e. to chase down CI test failures and identify the kind of queue/subscription being used just in case the issue does not repro with the randomly choses queue/subscription locally.

If we want to target a particular type when running locally, we can edit the test and use TestClientType enum directly

richardpark-msft replied:
It'd be nice to not have to push temp changes if we can avoid it. Is there a reason we're averse to having it pick the same set of tests on the same day?

HarshaNalluru commented:
> I wonder if we could say: CI runs do this random behavior but nightly CI runs do the full tests? Is that a possibility?

What's the purpose? We don't have any code path differentiating partitioned vs unpartitioned or queues vs subscriptions.

ramya-rao-a commented:
> I wonder if we could say: CI runs do this random behavior but nightly CI runs do the full tests? Is that a possibility?

Not at the moment unless we set up a third kind of CI other than the PR and nightly ones. Because we use the azp command to trigger the live tests from PR.

richardpark-msft commented:
> 
> 
> > I wonder if we could say: CI runs do this random behavior but nightly CI runs do the full tests? Is that a possibility?
> 
> Not at the moment unless we set up a third kind of CI other than the PR and nightly ones. Because we use the azp command to trigger the live tests from PR.

@KarishmaGhiya - is there a way to have a parameter or something (maybe an environment variable?) that only gets passed when the nightly tests run but wouldn't get passed when we run our normal live runs during CI?

richardpark-msft commented:
> 
> 
> > I wonder if we could say: CI runs do this random behavior but nightly CI runs do the full tests? Is that a possibility?
> 
> What's the purpose? We don't have any code path differentiating partitioned vs unpartitioned or queues vs subscriptions.

It's less about our code and just making sure that we our SDK doesn't have unexpected runtime issues with changes in Service Bus.

The bulk of our logic changes should be behind unit tests and I expect the live tests to shrink over time so this will become less of a burden and more of a sanity check.

ramya-rao-a commented:
/azp run js - servicebus - tests

ramya-rao-a commented:
/azp run js - servicebus - tests

ramya-rao-a commented:
/azp run js - servicebus - tests

ramya-rao-a commented:
/azp run js - servicebus - tests

ramya-rao-a commented:
/azp run js - servicebus - tests

ramya-rao-a commented:
Alright! This PR is ready for review.

Based on the OS, the gains are different.
Didnt really cut the time in half, because we did not have the repetition in ALL the places.

Before:

<img width="690" alt="Screen Shot 2020-07-13 at 5 05 16 PM" src="https://user-images.githubusercontent.com/16890566/87365433-0f595600-c52b-11ea-8537-eb257bee003a.png">

After:

<img width="681" alt="Screen Shot 2020-07-13 at 5 05 22 PM" src="https://user-images.githubusercontent.com/16890566/87365435-12544680-c52b-11ea-9890-fe53d3e43662.png">

ramya-rao-a commented:
Number of tests reduced from about 3500 to 2100!

ramya-rao-a commented:
Notes on change in this PR
- Added 3 helpers in test utils file to get a 
     - random test client type
     - random test client type with sessions enabled
     - random test client type with no sessions enabled
- All tests that used all 8 variations are updated to use 2 variations: sessions + no-sessions
- Tests that used hard-coded client type, but no variations are updated to use a random test client type
     - This is done only in files that got touched in the previous step
- Added some new tests in some places

HarshaNalluru commented on sdk/servicebus/service-bus/test/backupMessageSettlement.spec.ts:
```diff
@@ -45,503 +50,285 @@ describe("Backup message settlement - Through ManagementLink", () => {
+    if (entityNames.usesSessions) {
+      should.equal(errorWasThrown, true, "Error was not thrown for messages with session-id");
+    } else {
+      should.equal(errorWasThrown, false, "Error was thrown for sessions without session-id");
```
OOPS!
```suggestion
      should.equal(errorWasThrown, false, "Error was thrown for messages without session-id");
```

HarshaNalluru commented on sdk/servicebus/service-bus/test/backupMessageSettlement.spec.ts:
```diff
@@ -45,503 +50,285 @@ describe("Backup message settlement - Through ManagementLink", () => {
-    afterEach(async () => {
-      await afterEachTest();
-    });
+  it(noSessionTestClientType + ": complete() removes message", async function(): Promise<void> {
```
What happened to 
`[noSessionTestClientType, withSessionTestClientType].forEach((testClientType) => {` ?

ramya-rao-a replied:
I had that format at one point, but moved away from it to keep things simple.
It helps with debugging or running individual tests.

For example, if I had to run/debug 1 test, I would add the `.only`
If I use the format you are suggesting here, the same test would run twice.
Now, I have to update the array to have a single entity type in order to run just the one test.

So, making things easier for people to use the tests trumped the need to optimize the code itself.
